xwiki-commons-core/xwiki-commons-store/xwiki-commons-store-blob/xwiki-commons-store-blob-api/src/main/java/org/xwiki/store/blob/Blob.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.store.blob;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import org.xwiki.stability.Unstable;
+import org.xwiki.store.StreamProvider;
+
+/**
+ * A Blob is a piece of data stored in a BlobStore.
+ *
+ * @version $Id$
+ * @since 17.7.0RC1
+ */
+@Unstable
+public interface Blob extends StreamProvider
+{
+    /**
+     * @return the store where this blob is stored
+     */
+    BlobStore getStore();
+
+    /**
+     * @return the path of this blob inside its store
+     */
+    BlobPath getPath();
+
+    /**
+     * @return true if the blob exists, false otherwise
+     * @throws BlobStoreException when the existence of the blob cannot be determined
+     */
+    boolean exists() throws BlobStoreException;
+
+    /**
+     * Get the size of this blob.
+     *
+     * @return the size of this blob in bytes, or -1 if the blob doesn't exist
+     * @throws BlobStoreException when the size cannot be determined
+     */
+    long getSize() throws BlobStoreException;
+
+    /**
+     * @param conditions the conditions that must be satisfied before writing to this blob
+     * @return an OutputStream to write data to this blob
+     * @throws BlobStoreException if the blob cannot be written, for example because its name is invalid. There is no
+     *     guarantee that in such a case an exception will be thrown, the exception could also only be thrown when data
+     *     is written to the stream, or when the stream is closed.
+     */
+    OutputStream getOutputStream(WriteCondition... conditions) throws BlobStoreException;
+
+    /**
+     * Write the content of the given InputStream to this blob.
+     *
+     * @param inputStream the InputStream to read data from
+     * @param conditions the conditions that must be satisfied before writing to this blob
+     * @throws BlobStoreException if the InputStream cannot be read or the blob cannot be written, for example because
+     * its name is invalid.
+     */
+    void writeFromStream(InputStream inputStream, WriteCondition... conditions) throws BlobStoreException;
```

Comment:
You think there are cases when getOutputStream would not be enough or that passing the inputstream to write would allow some optimizations ? As otherwise, I don't see much point of exposing this API, even as a helper IOUtils#copy should be good enough.

Reply:
For now, I've added a comment to reconsider the method.

Reply:
I've just noticed that I still haven't resolved this. So the current status is:

* There is still no implementation that benefits from `writeFromStream`. I've decided against the async client as it would introduce a lot of overhead and boilerplate code for currently almost no benefit (except maybe for uploading large files).
* In the calling code, `writeFromStream` seems quite convenient and is used in a few places.

So I'm still not sure what to do here.